Stellar-transaction.x

```diff
@@ -530,12 +530,31 @@ case HOST_FUNCTION_TYPE_INSTALL_CONTRACT_CODE:
     InstallContractCodeArgs installContractCodeArgs;
 };
 
+struct AuthorizedInvocation
+{
+    Hash contractID;
+    SCSymbol functionName;
+    SCVec args;
+    AuthorizedInvocation subInvocations<>;
```

Comment:
Seems like with this change we're now storing the args for every invocation in the whole tree. Does that change behavior?

Reply:
It seems like this makes it harder for wallets to differentiate authorizedInvocations from incidental invocations?

Reply:
It does make it more strict for some edge cases (which is why I switched to this structure), but overall doesn't change much. Before if you had a call stack with authorized calls `A(a, b, c)->B(d, e, f)` you'd had to sign a vector of two call stacks: `[{A(a,b,c)}, {A->B(d, e, f)}]`. Now you just sign both in a tree, so both are equivalent. The edge cases come up when the same address signs multiple stacks and it could be possible to authorize unrelated calls, which may or may not be abusable. Enforcing the tree structure makes the rules much more consistent.
FWIW in both cases the 'tree' only corresponds to the authorized calls. Calls that don't need authorization may be inserted in between, so that things like proxies/routers etc. are possible without leaking into signatures.

Reply:
> It seems like this makes it harder for wallets to differentiate authorizedInvocations from incidental invocations?

What do you mean by 'incidental' invocations? As mentioned above (and as the name suggests) only authorized invocations have to be included.

The difference for wallets is walking a tree vs iterating a vector. I'd *hope* it's not a big implementation issue.

Reply:
I mean, how would a wallet select the "relevant" invocations to display to the user for approval?

Reply:
> In the case that only C.baz calls the verification, A.foo, and B.bar would be "incidental" because they do not call the verification. 

Again, in this case only the signature for C.baz is required. There are really no truly 'incidental' things to sign, because every call has requested a signature for some reason. I realize there can be some cases with redundancy (e.g,. consider signing something like `swap_with_some_dex->generic_swap->xfer`, the swap signature is likely redundant, but probably it would've been better if the contract writer just didn't require the `generic_swap` signature in the first place). Basically deep trees are not disallowed, but I wouldn't anticipate them to come up frequently.

In any case, I think that wallet should keep track of the relevant entities (tokens, nfts, something else) and highlight them no matter in which place in the tree they appear.

> We could keep the key structure and only retain topArgs for authorizations if args were optional SCVec *args;, but if keeping all args in the tree is more secure,

*In theory* we could omit in-between args even for authorized calls. I'm not sure if there are some attacks that are made possible by this. But if there are some, then we're in trouble, while OTOH I don't really anticipate deep trees to be prevalent, as mentioned above.

Reply:
I think I am misunderstanding:
> There are really no truly 'incidental' things to sign, because every call has requested a signature for some reason.

Isn't a signature only requested when a contract verifies the address (or whatever the method is called to consume the auth)?

Reply:
> relevant entities

Ok.. I guess I was thinking of using "consumes an authorizedInvocation" as a proxy for relevance

Reply:
Discussed elsewhere, but my misunderstanding was that I didn't realize there were multiple trees. And the nodes in the trees are only methods which call `require_auth`. Non-authenticating methods are skipped.

Reply:
Here is more formal definition for building the tree in the signature payload:

Consider the rooted directed sub-contract call tree created by a single top-level contract invocation. From the perspective of a single address there are two kinds of nodes in the tree. Let's call them "N-nodes" - the nodes that don't require auth for the address and "A-nodes" - the nodes that call `require_auth` for that address. Now we perform the following procudure to create the signature payload: remove all the edges and all the N-nodes from the tree (so we're left with only A-nodes). Then create directed edges between the A-nodes such that there is an edge from A1 to A2 if and only if there is a path from A1 to A2 in the original tree such that only passes through N-nodes (or no other nodes at all). 

Note, that this procedure may result in multiple trees, but this is more of an edge case (that would mean that the contract makes some unrelated calls). Multiple trees would be represented by multiple `ContractAuth` entries here.